CPLAT-12619 support new factory syntax (#660)
Workiva/over_react: doc/new_boilerplate_migration.md

* __[New Boilerplate Updates](#new-boilerplate-updates)__
    * [Design Constraints](#design-constraints)
    * [Updates](#updates)

>         * [Add `castUiFactory` Utility](#add-castuifactory-utility)

Reviewer: We should also add instructions for the codemod for these changes

### Updates

#### Add `castUiFactory` Utility

> A utility called `castUiFactory` has been added that protects against implicit cast warnings
> on factory declarations. All that needs to be done is to wrap the generated factory with

Reviewer: ```suggestion
A utility called `castUiFactory` has been added that prevent implicit cast errors 
(which are no longer ignorable as of Dart 2.9) on factory declarations. All that needs to be done is to wrap the generated factory with
```

`castUiFactory`, so that it can infer the typing from the left hand side and cast the

> factory (considered "dynamic") to the correct type.

Reviewer: ```suggestion
factory (considered "dynamic" before code generation is run) to the correct type.
```

```diff
@Factory()
UiFactory<FooProps> Foo =

> +    castUiFactory(_$Foo); // ignore: undefined_identifier

Reviewer: Missing the "before"
```suggestion
-    _$Foo; // ignore: undefined_identifier
+    castUiFactory(_$Foo); // ignore: undefined_identifier
```
